students/DanCornutt/session05/mailroom.py

```diff
@@ -0,0 +1,209 @@
+    Please give more next time.\n
+    \tLove,\n
+    \t\t\tThe Team""".format(donor=name, recent=DONORS[name][-1])
+    with open(name.replace(" ","") + "_thank_you.txt", 'w') as f_out:
+        f_out.write(text)
+
+
+def thank_you(donor="", all_users=False):
+    """Writes thank you letter to file, if all_users is True write letters for all users in database
+    """
+    if all_users:
+        for d in DONORS:
+            write_donor(d)
+    else:
+        donor = input("Type Donor Name =>").title()
+        if check_donor(donor):
+            write_donor(donor)
+        else:
+            print("Sorry I could not find the donor, exiting...")
+
+
+def donor_db():
+    """Editing Donor Menu"""
+    menu_selection(DATABASE_PROMPT, DATABASE_DISPATCH)
+
+
+def edit_donor():
+    answer = input("input donor name=> ").strip().title()
+    if answer:
+        add_donation(answer)
+    else:
+        print("The name must be letters only, returning...")
+        thank_you()
+
+def add_donation(donor_name):
+    """adds donation to donor records, adds donor if new donor
+
+    :param1: donar name
+    """
+    donation = add_money()
+    if donation:
+        DONORS.setdefault(donor_name, []).append(donation)
+        print(
+            "Thank you {} for your donation of ${:,.2f} dollars!".format(
+                donor_name, donation)
+        )
+
+
+def add_money():
+    """Takes input from user, validates donation
+
+    :returns: donation amount or None if donation is invalid.
+    """
+    donation = input("Please enter the donor amount larger than 0.01 => $")
+    try:
+        if float(donation) < 0.01:
+            raise ValueError
+    except ValueError:
+        print("Donation Error, numbers only. Value must be number greater than 0.01")
+        return None
+    else:
+        return float(donation)
+
+
+def thank_you_all():
+    """Writes thank you email to all users"""
+    thank_you(all_users=True)
+
+
+def report():
+    """Finds column widths, sorts donors based on amount donated.
+    :returns: report string
+    """
+    rpt_sheet = []
+    len_col = OrderedDict({
+        "n_size": len("Donor Name"),
+        "t_size" : len("Total Given"),
+        "ng_size": len("Num Gifts"),
+        "ag_size": len("Average Gift")
+    })
+
+    for d in DONORS.items():
+        if len(d[0]) > len_col["n_size"]:
+            len_col["n_size"] = len(d[0])
+        if len(str(sum(d[1]))) > len_col["t_size"]:
+            len_col["Tt_size"] = len(str(sum(d[1])))
+        if len((d[1])) > len_col["ng_size"]:
+            len_col["ng_size"] = len((d[1]))
+        if len(str(sum(d[1])/len(d[1]))) > len_col["ag_size"]:
+            len_col["ag_size"] = len(str(sum(d[1])/len(d[1])))
+        rpt_sheet.append((d[0], sum(d[1]), len(d[1]), sum(d[1])/len(d[1])))
+    rpt_sheet.sort(key=return_total, reverse=True)
+
+    sheet = (
+        "{nm:{mnm}} | {tot:<{mtot}} | {ng:<{mng}} | {ag:<{mag}}\n{header}".format(
```

Comment:
kind of a lot in one "line" of code -- maybe put all that data in a dict -- then pass it in to format with **